lnrpc/routerrpc/router_server_test.go

```diff
@@ -0,0 +1,169 @@
+		sentFromServer: make(chan *lnrpc.Payment, 10),
+	}
+}
+
+type StreamMock struct {
+	grpc.ServerStream
+	ctx            context.Context
+	sentFromServer chan *lnrpc.Payment
+}
+
+func makeControlTowerMock() *ControlTowerMock {
+	towerMock := &ControlTowerMock{
+		queue: queue.NewConcurrentQueue(20),
+	}
+	towerMock.queue.Start()
+
+	return towerMock
+}
+
+type ControlTowerMock struct {
+	queue *queue.ConcurrentQueue
+	routing.ControlTower
+}
+
+func (t *ControlTowerMock) SubscribeAllPayments() (
+	*routing.ControlTowerSubscriber, error) {
+
+	return &routing.ControlTowerSubscriber{
+		Updates: t.queue.ChanOut(),
+	}, nil
+}
+
+func (m *StreamMock) Context() context.Context {
+	return m.ctx
+}
+
+func (m *StreamMock) Send(p *lnrpc.Payment) error {
+	m.sentFromServer <- p
+	return nil
+}
+
+// TestTrackPaymentsInflightUpdate tests whether all updates from the control
+// tower are propagated to the client.
+func TestTrackPaymentsInflightUpdates(t *testing.T) {
+	// Setup mocks and request.
+	request := &TrackPaymentsRequest{
+		NoInflightUpdates: false,
+	}
+	towerMock := makeControlTowerMock()
+	stream := makeStreamMock()
+	server := &Server{
+		cfg: &Config{
+			RouterBackend: &RouterBackend{
+				Tower: towerMock,
+			},
+		},
+	}
+
+	// Listen to payment updates in a goroutine.
+	go func() {
+		err := server.TrackPayments(request, stream)
+		require.NoError(t, err)
+	}()
+
+	// Enqueue some payment updates on the mock.
+	towerMock.queue.ChanIn() <- &channeldb.MPPayment{
+		Info:   &channeldb.PaymentCreationInfo{},
+		Status: channeldb.StatusInFlight,
+	}
+	towerMock.queue.ChanIn() <- &channeldb.MPPayment{
+		Info:   &channeldb.PaymentCreationInfo{},
+		Status: channeldb.StatusSucceeded,
+	}
+
+	// Wait until there's 2 updates or the deadline is exceeded.
+	deadline := time.Now().Add(1 * time.Second)
+	for {
+		if len(stream.sentFromServer) == 2 {
+			break
+		}
+
+		if time.Now().After(deadline) {
+			require.FailNow(t, "deadline exceeded.")
+		}
+	}
+
+	// Both updates should be sent to the client.
+	require.Len(t, stream.sentFromServer, 2)
+
+	// The updates should be in the right order.
+	payment := <-stream.sentFromServer
+	require.Equal(t, lnrpc.Payment_IN_FLIGHT, payment.Status)
+	payment = <-stream.sentFromServer
+	require.Equal(t, lnrpc.Payment_SUCCEEDED, payment.Status)
+}
+
+// TestTrackPaymentsInflightUpdate tests whether only final updates from the
+// control tower are propagated to the client when noInflightUpdates = true.
+func TestTrackPaymentsNoInflightUpdates(t *testing.T) {
+	// Setup mocks and request.
+	request := &TrackPaymentsRequest{
+		NoInflightUpdates: true,
+	}
+	towerMock := &ControlTowerMock{
+		queue: queue.NewConcurrentQueue(20),
+	}
+	towerMock.queue.Start()
+	stream := makeStreamMock()
+	server := &Server{
+		cfg: &Config{
+			RouterBackend: &RouterBackend{
+				Tower: towerMock,
+			},
+		},
+	}
+
+	// Listen to payment updates in a goroutine.
+	go func() {
+		err := server.TrackPayments(request, stream)
```

Comment:
q: what happens to this go routine after the test is done?

Reply:
Well that was more work than I expected...

Indeed that go routine would last forever I think (until the test run was done probably)
So I decided to pass a cancellable context to that stream so the trackpayments call would complete after the test was complete.

Turns out that caused a panic, because the ControlTowerSubscriber tries to close the `quit` chan when the TrackPayments call completes. So I had to convert ControlTowerSubscriber to an interface in order to be able to mock it in the test.

Did these things in a newly added commit.

Reply:
Sorry for that :pray: